plugins/dbus_announce.c

```diff
@@ -0,0 +1,198 @@
+#include "system.h"
+
+#include <dbus/dbus.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <rpm/rpmlog.h>
+#include <rpm/rpmts.h>
+#include <rpm/rpmdb.h>
+#include "lib/rpmplugin.h"
+
+struct dbus_announce_data {
+    DBusConnection * bus;
+    int logging;
+    int old_dboffset;
+};
+
+static rpmRC dbus_announce_init(rpmPlugin plugin, rpmts ts)
+{
+    struct dbus_announce_data * state = rcalloc(1, sizeof(*state));
+    rpmPluginSetData(plugin, state);
+    return RPMRC_OK;
+}
+
+static rpmRC open_dbus(rpmPlugin plugin, rpmts ts)
+{
+    struct stat st;
+    DBusError err;
+    int rc = 0;
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+
+    /* Assume we are logging but... */
+    state->logging = 1;
+
+    /* ...don't log test transactions */
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST|RPMTRANS_FLAG_BUILD_PROBS))
+	state->logging = 0;
+
+    /* ...don't log chroot transactions */
+    if (!rstreq(rpmtsRootDir(ts), "/"))
+	state->logging = 0;
+
+    /* Don't open */
+    if (!state->logging || state->bus)
+	return RPMRC_OK;
+
+    if (lstat("/run/systemd/system/", &st) == 0) {
+        if (S_ISDIR(st.st_mode)) {
```

Comment:
Pardon my ignorance, but why is this depending on some systemd directory? Does dbus need systemd?

Reply:
DBus the protocol doesn't need systemd (it predates it) but lately integrates with it well.  The reason I think we do this check is it's a proxy for "is this system booted".

For example, there's no dbus running in a default podman/docker style container (neither is there systemd, or in fact usually there isn't any other processes at all except e.g. yum/zypper running).  So there's no reason to try announcing.

Related discussion and code: https://github.com/systemd/systemd/pull/7631

Reply:
That check looks like it's "inherited" from systemd_inhibit plugin, where this is relevant because we're talking to systemd over dbus (see commit ec6495d79fa9834bca3394f56e9f6eb933030b1d). This is an entirely different case though, other processes might want to know about rpmdb state changes regardless of whether systemd is running or not.

Reply:
Fixed.